universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,198 @@
+    return m_value_refs.end();
+}
+
+NamedValueRefManager& NamedValueRefManager::GetNamedValueRefManager() {
+    ErrorLogger() << "NamedValueRefManager::GetNamedValueRefManager starts in process " << ::getpid() << " thread: " << std::this_thread::get_id();
+    static NamedValueRefManager manager; // function local 
+    ErrorLogger() << "NamedValueRefManager::GetNamedValueRefManager at " << &manager << " in process " << ::getpid() << " thread: " << std::this_thread::get_id();
+    return manager;
+}
+
+
+unsigned int NamedValueRefManager::GetCheckSum() const {
+    CheckPendingRefs();
+    unsigned int retval{0};
+    /*    for (auto const& name_type_pair : m_value_refs)
+        CheckSums::CheckSumCombine(retval, name_type_pair);  // TODO FIXME
+    CheckSums::CheckSumCombine(retval, m_value_refs.size());
+
+
+    DebugLogger() << "NamedValueRefManager checksum: " << retval;*/
+    return retval;
+}
+
+
+template <typename T>
+void NamedValueRefManager::RegisterValueRef(std::string valueref_name, std::unique_ptr<T> vref) {
+    InfoLogger() << "Register valueref for " << valueref_name << ": " << vref->Description();
+    if (m_value_refs.count(valueref_name)>0) {
+        ErrorLogger() << "Skip registration for already registered valueref for " << valueref_name;
+        ErrorLogger() << "Number of registered ValueRefs: " << m_value_refs.size();
+        return;
+    }
+    m_value_refs.emplace(valueref_name, std::move(std::unique_ptr<ValueRef::ValueRefBase>(std::move(vref))));
+    ErrorLogger() << "Number of registered ValueRefs: " << m_value_refs.size();
+}
+
+
+NamedValueRefManager& GetNamedValueRefManager()
+{ return NamedValueRefManager::GetNamedValueRefManager(); }
+
+ValueRef::ValueRefBase* GetValueRefBase(const std::string& name)
+{
+    InfoLogger() << "NamedValueRefManager::GetValueRefBase look for registered valueref for \"" << name << '"';
+    if ( GetNamedValueRefManager().GetValueRefBase(name) )
+        return GetNamedValueRefManager().GetValueRefBase(name);
+    InfoLogger() << "NamedValueRefManager::GetValueRefBase could not find registered valueref for \"" << name << '"';
+    return nullptr;
+}
+
+template <typename T>
+ValueRef::ValueRef<T>* GetValueRef(const std::string& name)
+{ return GetNamedValueRefManager().GetValueRef<T>(name); }
+
+template <>
+ValueRef::ValueRef<int>* GetValueRef(const std::string& name)
+{ return GetNamedValueRefManager().GetValueRef<int>(name); }
+
+template <>
+ValueRef::ValueRef<double>* GetValueRef(const std::string& name)
+{ return GetNamedValueRefManager().GetValueRef<double>(name); }
+
+template <typename T>
+void RegisterValueRef(std::string name, std::unique_ptr<T>&& vref) {
+    return GetNamedValueRefManager().RegisterValueRef<T>(name, std::move(vref));
+}
+
+template <>
+void RegisterValueRef(std::string name, std::unique_ptr<ValueRef::ValueRef<int>>&& vref) {
+    return GetNamedValueRefManager().RegisterValueRef<ValueRef::ValueRef<int>>(name, std::move(vref));
```

Comment:
std::move(name)

Reply:
so `GetNamedValueRefManager::RegisterValueRef(std::string&& name, ...)`, yes?

Reply:
if it compiles, but then you can't pass in an lvalue, which I think you usually do... so would need more `move` calls...

by value is more general

just `std::string&` is possibly OK... depends what the calling code does...

Reply:
std::string& is suboptimal, i would have to do a copy inside anyway. 
went for the rvalue - this is tightly coupled and should be accessed through the global function which takes a old school lvalue.